consumer_no_netlink.c

```diff
@@ -0,0 +1,299 @@
+ *
+ * Using TCMU without netlink is useful for scenarios where they may be
+ * multiple applications using TCMU on one host. This sort of configuration is
+ * not possible when TCMU uses netlink because backstore device events are
+ * broadcasted over netlink to all TCMU subscribers.  This mean one application
+ * can receive netlink events for devices which pertain to another application.
+ * If an app receives a netlink event for a device it does not know about libtcmu
+ * will return an error message and the kernel will fail the device action.
+ *
+ * The example creates a TCMU backing store (using the create_backstore.sh
+ * script) and then notifies libtcmu that the backstore has been created using
+ * tcmulib_notify_device_added. The example then sits in a loop processing
+ * commands.  Note: unless you set up an HBA and LUN there won't be any
+ * commands to process.  When the app recieves a termination signal it exits
+ * the processing loop. It then notifies libtcmu that the backstore is about to
+ * be removed via the tcmulib_notify_device_removed method. We then delete the
+ * backstore using the remove_backstore.sh script.
+ *
+ * This example must be ran in the the same directory as its accompanying
+ * bash scripts, i.e. the top level repo directory.
+ *
+ * NOTE: TCMU without netlink is not a panacea to the multi-TCMU problem. One
+ * still has to be careful to ensure that multiple apps don't try and create
+ * the same backstore (user_XXXX) at the same time. It may be necessary to
+ * synchronise/serialise other configs actions too.
+ *
+ * NOTE: this is a bare-bones example! It eschews any proper error handling,
+ * uses hardcoded values, and relies on bash scripts. It's intended to give
+ * a rough outline of how to use the feature and should not be used as the
+ * basis of any production code.
+ */
+
+#define _GNU_SOURCE
+#include <endian.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <poll.h>
+#include <scsi/scsi.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "libtcmu.h"
+#include "libtcmu_log.h"
+#include "scsi.h"
+#include "scsi_defs.h"
+#include "target_core_user_local.h"
+
+#define LOG_DIR "/var/log"
+#define CFG_STRING "tcm-user/1/test/foo/xxx"
+
+struct tcmu_device *tcmu_dev_array[128];
+size_t dev_array_len = 0;
+
+struct foo_state {
+	int fd;
+	uint64_t num_lbas;
+	uint32_t block_size;
+};
+
+static volatile int keep_running = 1;
+
+static void signal_handler(int) {
+	keep_running = 0;
+}
+
+static bool run_command(char* buffer, size_t size, char* cmd)
+{
+	FILE *f = popen(cmd, "r");
+	if (!f) {
+		tcmu_err("couldn't run command %s\n", cmd);
+		return false;
+	}
+
+	fgets(buffer, size, f);
+	pclose(f);
+	return true;
+}
+
+static bool create_backstore()
+{
+	int ret = system("./create_backstore.sh");
```

Comment:
This seems incorrect approach.
Why not comply to the current `consumer.c`, which will create the devices via targetcli or ceph-iscsi ?

Reply:
For one, I don't believe `targetcli` has the ability to specify `nl_reply_supported=-1` when creating the backstore. This is vital if you want to use TCMU without netlink. Secondly, I think there's merit in having a completely stand-alone example which doesn't require the user to have external tools installed.